test/common/upstream/eds_subscription_factory_test.cc

```diff
@@ -0,0 +1,108 @@
+  EdsSubscriptionFactory factory;
+
+  EXPECT_CALL(dispatcher_, createTimer_(_)).Times(2);
+  envoy::api::v2::core::ConfigSource first_config;
+  first_config.mutable_api_config_source()->add_grpc_services()->mutable_envoy_grpc()->set_cluster_name("first_cluster");
+  Config::GrpcMux& first_mux = factory.getOrCreateMux(local_info_, std::unique_ptr<Grpc::MockAsyncClient>(new Grpc::MockAsyncClient()), dispatcher_, method_descriptor_,
+                                                      random_, first_config.api_config_source(), stats_store_, rate_limit_settings_);
+  envoy::api::v2::core::ConfigSource second_config;
+  second_config.mutable_api_config_source()->add_grpc_services()->mutable_envoy_grpc()->set_cluster_name("second_cluster");
+  Config::GrpcMux& second_mux = factory.getOrCreateMux(local_info_, std::unique_ptr<Grpc::MockAsyncClient>(new Grpc::MockAsyncClient()), dispatcher_, method_descriptor_,
+                                                       random_, second_config.api_config_source(), stats_store_, rate_limit_settings_);
+  ASSERT_TRUE(&first_mux != &second_mux);
+}
+
+TEST_F(EdsSubscriptionFactoryTest, ShouldUseGetOrCreateMuxWhenApiConfigSourceIsUsed) {
+  EdsSubscriptionFactory factory;
+
+  envoy::api::v2::core::ConfigSource config;
+  config.mutable_api_config_source()->set_api_type(envoy::api::v2::core::ApiConfigSource::GRPC);
+  config.mutable_api_config_source()->add_grpc_services()->mutable_envoy_grpc()->set_cluster_name("first_cluster");
+
+  NiceMock<Grpc::MockAsyncClientFactory>* async_client_factory{new NiceMock<Grpc::MockAsyncClientFactory>()};
+  EXPECT_CALL(*async_client_factory, create()).WillOnce(Invoke([] { return std::make_unique<NiceMock<Grpc::MockAsyncClient>>(); }));
+
+  NiceMock<Grpc::MockAsyncClientManager> async_client_manager;
+  EXPECT_CALL(async_client_manager, factoryForGrpcService(_, _, _))
+      .WillOnce(Invoke([async_client_factory] (const envoy::api::v2::core::GrpcService&, Stats::Scope&, bool) {
+        return Grpc::AsyncClientFactoryPtr(async_client_factory); }));
+  EXPECT_CALL(cm_, grpcAsyncClientManager()).WillOnce(ReturnRef(async_client_manager));
+  EXPECT_CALL(dispatcher_, createTimer_(_));
+
+  auto subscription = factory.subscriptionFromConfigSource(config, local_info_, dispatcher_, cm_, random_, stats_store_, nullptr,
+    "envoy.api.v2.EndpointDiscoveryService.FetchEndpoints", "envoy.api.v2.EndpointDiscoveryService.StreamEndpoints");
+
+  Config::GrpcMux& expected_mux = factory.getOrCreateMux(local_info_, std::unique_ptr<Grpc::MockAsyncClient>(new Grpc::MockAsyncClient()), dispatcher_, method_descriptor_,
+                                                     random_, config.api_config_source(), stats_store_, rate_limit_settings_);
+
+  Config::Subscription<envoy::api::v2::ClusterLoadAssignment>& eds_subscription = *subscription;
+  ASSERT_TRUE(&expected_mux == &(dynamic_cast<Config::EdsGrpcSubscriptionImpl&>(eds_subscription).grpcMux()));
```

Comment:
I think you are after `EXPECT` in all these final test statements, and specifically `EXPECT_EQ` here.

Reply:
done.